src/cdomains/valueDomain.ml

```diff
@@ -562,11 +562,12 @@ struct
     | (x, `Bot) -> x
     | (`Int x, `Int y) -> (try `Int (ID.widen x y) with IntDomain.IncompatibleIKinds m -> Messages.warn ~category:Analyzer "%s" m; `Top)
     | (`Float x, `Float y) -> `Float (FD.widen x y)
+    (* TODO: symmetric widen, wtf? *)
     | (`Int x, `Address y)
     | (`Address y, `Int x) -> `Address (match ID.to_int x with
-        | Some x when BI.equal BI.zero x -> AD.widen AD.null_ptr y
-        | Some x -> AD.(widen y not_null)
-        | None -> AD.widen y AD.top_ptr)
+        | Some x when BI.equal BI.zero x -> AD.widen AD.null_ptr (AD.join AD.null_ptr y)
+        | Some x -> AD.(widen y (join y not_null))
+        | None -> AD.widen y (AD.join y AD.top_ptr))
```

Comment:
Is this necessary here, or will not the solver already ensure that this join has happened?

Reply:
I think the issue stems from the fact that `widen` seems to be copied and modified from `join`, where the `Int`-`Address` (and vice versa) cases are handled symmetrically together. Widening is not commutative, so the misuse of `AD.widen` probably only happens from one of the cases.

I suppose the proper solution would be to remove the symmetry and handle both cases separately (and appropriately), in which case it might no longer be necessary, but I didn't want to make such delicate changes to these domains here.